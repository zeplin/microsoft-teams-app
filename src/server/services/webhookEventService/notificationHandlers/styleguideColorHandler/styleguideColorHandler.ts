import {
    WebhookEvent,
    StyleguideColorCreateEvent,
    StyleguideColorUpdateEvent
} from "../../../../adapters/zeplin/types";
import { NotificationHandler } from "../NotificationHandler";
import { SHORT_DELAY } from "../constants";
import { commonTeamsCard, MessageCard } from "../teamsCardTemplates";
import { md } from "../md";
import { getRandomEmoji } from "../getRandomEmoji";
import { getRedirectURLForZeplinApp, getWebAppURL } from "../zeplinURL";
import { getColorUpdateMessage } from "../getStyleUpdateMessage";

type Event = StyleguideColorCreateEvent | StyleguideColorUpdateEvent;

class StyleguideColorHandler extends NotificationHandler<Event> {
    delay = SHORT_DELAY;
    private getText(events: Event[]): string {
        const [{
            payload: {
                action,
                context: {
                    styleguide: {
                        name: styleguideName
                    }
                },
                resource: {
                    data: {
                        name: pivotColorName
                    }
                }
            }
        }] = events;
        const actionText = action === "created" ? "added to" : "updated in";
        return events.length === 1
            ? md`**${pivotColorName}** is ${actionText} _${styleguideName}_! ${getRandomEmoji()}`
            : md`**${events.length} colors** are ${actionText} _${styleguideName}_! ${getRandomEmoji()}`;
    }

    private getSectionText(events: Event[]): string {
        const [{
            payload: {
                context: {
                    styleguide: {
                        platform: styleguidePlatform
                    }
                }
            }
        }] = events;

        return getColorUpdateMessage(styleguidePlatform);
    }

    private getWebappURL(
        events: Event[]
    ): string {
        const [{
            payload: {
                context: {
                    styleguide: {
                        id: styleguideId
                    }
                }
            }
        }] = events;
        const pathname = `styleguide/${styleguideId}/colors`;
        const searchParams = {
            cid: events.map(event => event.payload.resource.id)
        };

        return getWebAppURL(pathname, searchParams);
    }

    private getZeplinAppURI(
        events: Event[]
    ): string {
        const [{
            payload: {
                context: {
                    styleguide: {
                        id: styleguideId
                    }
                }
            }
        }] = events;
        const searchParams = {
            stid: styleguideId,
            cids: events.map(event => event.payload.resource.id)
        };

        return getRedirectURLForZeplinApp("colors", searchParams);
    }

    shouldHandleEvent(event: WebhookEvent): event is Event {
        return event.payload.action !== "deleted";
    }

    getTeamsMessage(
        events: Event[]
    ): MessageCard {
        return commonTeamsCard({
            text: this.getText(events),
            section: {
                text: this.getSectionText(events)
            },
            links: [{
                title: "Open in App",
                url: this.getZeplinAppURI(events)
            }, {
                title: "Open in Web",
                url: this.getWebappURL(events)
            }]
        });
    }
}

export const styleguideColorHandler = new StyleguideColorHandler();
