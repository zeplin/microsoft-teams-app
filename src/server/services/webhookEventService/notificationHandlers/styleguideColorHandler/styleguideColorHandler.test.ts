import { styleguideColorHandler } from "./styleguideColorHandler";
import {
    StyleguideColorCreateEvent,
    StyleguideColorUpdateEvent,
    StyleguidePlatform
} from "../../../../adapters/zeplin/types";

type GetDummyEventParams = {
    action?: string;
    colorId?: string;
    colorName?: string;
    styleguidePlatform?: string;
}

function getDummyEvent({
    action = "created",
    colorId = "colorId",
    colorName = "colorName",
    styleguidePlatform = StyleguidePlatform.WEB
}: GetDummyEventParams = {}): StyleguideColorCreateEvent | StyleguideColorUpdateEvent {
    return {
        payload: {
            action,
            context: {
                styleguide: {
                    id: "styleguideId",
                    name: "styleguideName",
                    platform: styleguidePlatform
                }
            },
            resource: {
                id: colorId,
                data: {
                    id: colorId,
                    name: colorName
                }
            }
        }
    } as StyleguideColorCreateEvent | StyleguideColorUpdateEvent;
}

describe("styleguideColorHandler", () => {
    describe("getTeamsMessage method", () => {
        beforeAll(() => {
            jest.spyOn(Math, "random").mockReturnValue(0);
        });

        describe("for created notification", () => {
            it("should match snapshot when there is only 1 event", () => {
                expect(
                    styleguideColorHandler.getTeamsMessage([getDummyEvent()])
                ).toMatchSnapshot();
            });

            it("should match snapshot when there are more than 1 events", () => {
                expect(
                    styleguideColorHandler.getTeamsMessage([
                        getDummyEvent(),
                        getDummyEvent({
                            colorId: "anotherColorId",
                            colorName: "anotherColorName"
                        })
                    ])
                ).toMatchSnapshot();
            });

            it.each([
                StyleguidePlatform.ANDROID,
                StyleguidePlatform.IOS,
                StyleguidePlatform.MAC_OS,
                StyleguidePlatform.WEB,
                StyleguidePlatform.BASE
            ])("should match snapshot for every possible styleguide platform",
                styleguidePlatform => {
                    expect(
                        styleguideColorHandler.getTeamsMessage([getDummyEvent({ styleguidePlatform })])
                    ).toMatchSnapshot();
                }
            );
        });

        describe("for updated notification", () => {
            it("should match snapshot when there is only 1 event", () => {
                expect(
                    styleguideColorHandler.getTeamsMessage([getDummyEvent({ action: "updated" })])
                ).toMatchSnapshot();
            });

            it("should match snapshot when there are more than 1 events", () => {
                expect(
                    styleguideColorHandler.getTeamsMessage([
                        getDummyEvent({
                            action: "updated"
                        }),
                        getDummyEvent({
                            action: "updated",
                            colorId: "anotherColorId",
                            colorName: "anotherColorName"
                        })
                    ])
                ).toMatchSnapshot();
            });

            it.each([
                StyleguidePlatform.ANDROID,
                StyleguidePlatform.IOS,
                StyleguidePlatform.MAC_OS,
                StyleguidePlatform.WEB,
                StyleguidePlatform.BASE
            ])("should match snapshot for every possible styleguide platform",
                styleguidePlatform => {
                    expect(
                        styleguideColorHandler.getTeamsMessage([getDummyEvent({ styleguidePlatform })])
                    ).toMatchSnapshot();
                }
            );
        });
    });
});
