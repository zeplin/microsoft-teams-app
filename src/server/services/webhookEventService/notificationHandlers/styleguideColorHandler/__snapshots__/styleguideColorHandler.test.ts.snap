// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styleguideColorHandler getTeamsMessage method for created notification should match snapshot for every possible styleguide platform 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**colorName** is added to _styleguideName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure \`colors.xml\` is up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcolors%3Fstid%3DstyleguideId%26cids%3DcolorId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/styleguide/styleguideId/colors?cid=colorId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`styleguideColorHandler getTeamsMessage method for created notification should match snapshot for every possible styleguide platform 2`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**colorName** is added to _styleguideName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure \`UIColor\` category / extension is up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcolors%3Fstid%3DstyleguideId%26cids%3DcolorId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/styleguide/styleguideId/colors?cid=colorId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`styleguideColorHandler getTeamsMessage method for created notification should match snapshot for every possible styleguide platform 3`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**colorName** is added to _styleguideName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure \`NSColor\` category / extension is up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcolors%3Fstid%3DstyleguideId%26cids%3DcolorId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/styleguide/styleguideId/colors?cid=colorId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`styleguideColorHandler getTeamsMessage method for created notification should match snapshot for every possible styleguide platform 4`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**colorName** is added to _styleguideName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcolors%3Fstid%3DstyleguideId%26cids%3DcolorId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/styleguide/styleguideId/colors?cid=colorId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`styleguideColorHandler getTeamsMessage method for created notification should match snapshot for every possible styleguide platform 5`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**colorName** is added to _styleguideName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your styles are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcolors%3Fstid%3DstyleguideId%26cids%3DcolorId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/styleguide/styleguideId/colors?cid=colorId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`styleguideColorHandler getTeamsMessage method for created notification should match snapshot when there are more than 1 events 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**2 colors** are added to _styleguideName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcolors%3Fstid%3DstyleguideId%26cids%3DcolorId%2CanotherColorId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/styleguide/styleguideId/colors?cid=colorId&cid=anotherColorId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`styleguideColorHandler getTeamsMessage method for created notification should match snapshot when there is only 1 event 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**colorName** is added to _styleguideName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcolors%3Fstid%3DstyleguideId%26cids%3DcolorId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/styleguide/styleguideId/colors?cid=colorId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`styleguideColorHandler getTeamsMessage method for updated notification should match snapshot for every possible styleguide platform 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**colorName** is added to _styleguideName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure \`colors.xml\` is up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcolors%3Fstid%3DstyleguideId%26cids%3DcolorId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/styleguide/styleguideId/colors?cid=colorId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`styleguideColorHandler getTeamsMessage method for updated notification should match snapshot for every possible styleguide platform 2`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**colorName** is added to _styleguideName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure \`UIColor\` category / extension is up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcolors%3Fstid%3DstyleguideId%26cids%3DcolorId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/styleguide/styleguideId/colors?cid=colorId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`styleguideColorHandler getTeamsMessage method for updated notification should match snapshot for every possible styleguide platform 3`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**colorName** is added to _styleguideName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure \`NSColor\` category / extension is up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcolors%3Fstid%3DstyleguideId%26cids%3DcolorId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/styleguide/styleguideId/colors?cid=colorId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`styleguideColorHandler getTeamsMessage method for updated notification should match snapshot for every possible styleguide platform 4`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**colorName** is added to _styleguideName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcolors%3Fstid%3DstyleguideId%26cids%3DcolorId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/styleguide/styleguideId/colors?cid=colorId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`styleguideColorHandler getTeamsMessage method for updated notification should match snapshot for every possible styleguide platform 5`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**colorName** is added to _styleguideName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your styles are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcolors%3Fstid%3DstyleguideId%26cids%3DcolorId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/styleguide/styleguideId/colors?cid=colorId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`styleguideColorHandler getTeamsMessage method for updated notification should match snapshot when there are more than 1 events 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**2 colors** are updated in _styleguideName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcolors%3Fstid%3DstyleguideId%26cids%3DcolorId%2CanotherColorId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/styleguide/styleguideId/colors?cid=colorId&cid=anotherColorId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`styleguideColorHandler getTeamsMessage method for updated notification should match snapshot when there is only 1 event 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**colorName** is updated in _styleguideName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcolors%3Fstid%3DstyleguideId%26cids%3DcolorId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/styleguide/styleguideId/colors?cid=colorId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;
